File: ios/Cove/NodeScannerManager.swift
import os
import SwiftUI

extension WeakReconciler: NodeScannerReconciler where Reconciler == NodeScannerManager {}

private enum NodeScannerManagerError: LocalizedError {
    case closed

    var errorDescription: String? {
        "Node scanner manager is closed"
    }
}

@Observable final class NodeScannerManager: ReconcilingManager, NodeScannerReconciler {
    typealias Message = NodeScannerReconcileMessage

    private struct RustState {
        var rust: RustNodeScannerManager?
        var isClosed = false
    }

    private let logger = Log(id: "NodeScannerManager")
    @ObservationIgnored
    private let rustState = OSAllocatedUnfairLock(initialState: RustState())
    @ObservationIgnored
    private let rustBridge = DispatchQueue(
        label: "cove.NodeScannerManager.rustbridge",
        qos: .userInitiated
    )
    @ObservationIgnored
    private var pendingRustOperation: Task<Void, Never>?

    private(set) var state: NodeScannerState = .idle

    private var rust: RustNodeScannerManager? {
        rustState.withLock { $0.rust }
    }

    init() {
        let rust = RustNodeScannerManager()
        rustState.withLock { $0.rust = rust }
        rust.listenForUpdates(reconciler: WeakReconciler(self))
        state = rust.state()
    }

    deinit {
        close()
    }

    func start() async {
        try? await performRustOperation { rust in
            await rust.startScan()
        }
    }

    func stop() async {
        try? await performRustOperation { rust in
            await rust.stopScan()
        }
    }

    func createNode(_ found: FoundNode) async throws -> Node {
        try await performRustOperation { rust in
            try await rust.createNode(found: found)
        }
    }

    func trustAndCreateNode(_ candidate: TrustRequiredEndpoint) async throws -> Node {
        try await performRustOperation { rust in
            try await rust.trustAndCreateNode(candidate: candidate)
        }
    }

    func close() {
        guard takeRustForClose() != nil else { return }

        logger.debug("Closing NodeScannerManager")
    }

    private func performRustOperation<T: Sendable>(
        _ operation: @escaping @Sendable (RustNodeScannerManager) async throws -> T
    ) async throws -> T {
        try await withCheckedThrowingContinuation { continuation in
            rustBridge.async { [weak self] in
                guard let self else {
                    continuation.resume(throwing: NodeScannerManagerError.closed)
                    return
                }

                let previous = self.pendingRustOperation
                let task = Task { [weak self] in
                    await previous?.value

                    guard let self, let rust = self.rust else {
                        continuation.resume(throwing: NodeScannerManagerError.closed)
                        return
                    }

                    do {
                        try await continuation.resume(returning: operation(rust))
                    } catch {
                        continuation.resume(throwing: error)
                    }
                }
                self.pendingRustOperation = task
            }
        }
    }

    private func takeRustForClose() -> RustNodeScannerManager? {
        rustState.withLock { state in
            guard !state.isClosed else { return nil }

            state.isClosed = true
            let rust = state.rust
            state.rust = nil
            return rust
        }
    }

    var canApplyReconcileMessages: Bool {
        rust != nil
    }

    func apply(_ message: Message) {
        switch message {
        case let .scanStarted(generation, network, progress):
            guard currentGeneration.map({ generation > $0 }) ?? true else { return }

            state = .scanning(
                generation: generation,
                network: network,
                progress: progress,
                results: []
            )

        case let .scanProgress(generation, progress):
            guard case let .scanning(currentGeneration, network, _, results) = state,
                  currentGeneration == generation
            else {
                return
            }

            state = .scanning(
                generation: generation,
                network: network,
                progress: progress,
                results: results
            )

        case let .resultUpserted(generation, result):
            guard let state = updatingResults(for: generation, with: { results in
                results = upserting(result, into: results)
            }) else { return }

            self.state = state

        case let .resultRemoved(generation, ip):
            guard let state = updatingResults(for: generation, with: { results in
                results.removeAll { resultIp($0) == ip }
            }) else { return }

            self.state = state

        case let .scanStopped(generation):
            guard case let .scanning(currentGeneration, network, _, results) = state,
                  currentGeneration == generation
            else {
                return
            }

            state = .stopped(generation: generation, network: network, results: results)

        case let .scanFinished(generation):
            guard case let .scanning(currentGeneration, network, _, results) = state,
                  currentGeneration == generation
            else {
                return
            }

            state = .finished(generation: generation, network: network, results: results)

        case let .scanFailed(generation, error):
            guard case let .scanning(currentGeneration, network, _, results) = state,
                  currentGeneration == generation
            else {
                return
            }

            state = .failed(
                generation: generation,
                network: network,
                error: error,
                results: results
            )
        }
    }

    func logReconcile(message: Message) {
        logger.debug("reconcile: \(message)")
    }

    func logReconcileMany(messages: [Message]) {
        logger.debug("reconcile_messages: \(messages)")
    }

    private var currentGeneration: UInt64? {
        switch state {
        case .idle:
            nil
        case let .scanning(generation, _, _, _),
             let .stopped(generation, _, _),
             let .finished(generation, _, _),
             let .failed(generation, _, _, _):
            generation
        }
    }

    private func upserting(
        _ result: NodeScannerResult,
        into results: [NodeScannerResult]
    ) -> [NodeScannerResult] {
        var results = results

        if let index = results.firstIndex(where: { resultIp($0) == resultIp(result) }) {
            results[index] = result
        } else {
            results.append(result)
        }

        return results
    }

    private func updatingResults(
        for generation: UInt64,
        with update: (inout [NodeScannerResult]) -> Void
    ) -> NodeScannerState? {
        guard currentGeneration == generation else { return nil }

        switch state {
        case .idle, .failed:
            return nil

        case let .scanning(_, network, progress, currentResults):
            var results = currentResults
            update(&results)
            return .scanning(
                generation: generation,
                network: network,
                progress: progress,
                results: results
            )

        case let .stopped(_, network, currentResults):
            var results = currentResults
            update(&results)
            return .stopped(generation: generation, network: network, results: results)

        case let .finished(_, network, currentResults):
            var results = currentResults
            update(&results)
            return .finished(generation: generation, network: network, results: results)
        }
    }

    private func resultIp(_ result: NodeScannerResult) -> String {
        switch result {
        case let .verified(node):
            node.ip
        case let .trustRequired(endpoint):
            endpoint.ip
        }
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/NodeScannerManager.kt
package org.bitcoinppl.cove

import android.util.Log
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.bitcoinppl.cove_core.FoundNode
import org.bitcoinppl.cove_core.Node
import org.bitcoinppl.cove_core.NodeScannerReconcileMessage
import org.bitcoinppl.cove_core.NodeScannerReconciler
import org.bitcoinppl.cove_core.NodeScannerResult
import org.bitcoinppl.cove_core.NodeScannerState
import org.bitcoinppl.cove_core.RustNodeScannerManager
import org.bitcoinppl.cove_core.TrustRequiredEndpoint
import java.io.Closeable
import java.util.concurrent.atomic.AtomicBoolean

/** Owns one screen-local local-node discovery lifecycle */
// the suspend wrappers do suspend through the generated bindings; detekt cannot
// resolve them because the generated module fails its type analysis
@Suppress("TooManyFunctions", "RedundantSuspendModifier")
@Stable
class NodeScannerManager internal constructor(
    private val rust: RustNodeScannerManager,
) : NodeScannerReconciler,
    Closeable {
    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val isClosed = AtomicBoolean(false)
    private val rustGuard =
        RustHandleGuard(
            ownerName = "NodeScannerManager",
            handleName = "RustNodeScannerManager",
            isClosed = isClosed,
        ) {
            Log.w(TAG, it)
        }

    var state by mutableStateOf<NodeScannerState>(NodeScannerState.Idle)
        private set

    constructor() : this(RustNodeScannerManager())

    init {
        withRust {
            listenForUpdates(this@NodeScannerManager)
        }
        state = withRust { state() }
    }

    private fun <T> withRust(block: RustNodeScannerManager.() -> T): T =
        rustGuard.withHandle(rust, block)

    private suspend fun <T> withRustSuspend(block: suspend RustNodeScannerManager.() -> T): T =
        rustGuard.withHandleSuspend(rust, block)

    private suspend fun <T> withRustOrSuspend(
        defaultValue: T,
        block: suspend RustNodeScannerManager.() -> T,
    ): T = rustGuard.withHandleOrSuspend(rust, defaultValue, block)

    suspend fun start() {
        withRustOrSuspend(Unit) {
            startScan()
        }
    }

    suspend fun stop() {
        withRustOrSuspend(Unit) {
            stopScan()
        }
    }

    suspend fun createNode(found: FoundNode): Node =
        withRustSuspend {
            createNode(found)
        }

    suspend fun trustAndCreateNode(candidate: TrustRequiredEndpoint): Node =
        withRustSuspend {
            trustAndCreateNode(candidate)
        }

    private fun apply(message: NodeScannerReconcileMessage) {
        when (message) {
            is NodeScannerReconcileMessage.ScanStarted -> applyScanStarted(message)
            is NodeScannerReconcileMessage.ScanProgress -> applyScanProgress(message)
            is NodeScannerReconcileMessage.ResultUpserted -> applyResultUpserted(message)
            is NodeScannerReconcileMessage.ResultRemoved -> applyResultRemoved(message)
            is NodeScannerReconcileMessage.ScanStopped -> applyScanStopped(message)
            is NodeScannerReconcileMessage.ScanFinished -> applyScanFinished(message)
            is NodeScannerReconcileMessage.ScanFailed -> applyScanFailed(message)
        }
    }

    private fun applyScanStarted(message: NodeScannerReconcileMessage.ScanStarted) {
        val currentGeneration = currentGeneration
        if (currentGeneration != null && message.generation <= currentGeneration) return

        state =
            NodeScannerState.Scanning(
                generation = message.generation,
                network = message.network,
                progress = message.progress,
                results = emptyList(),
            )
    }

    private fun applyScanProgress(message: NodeScannerReconcileMessage.ScanProgress) {
        val current = scanningState(message.generation) ?: return
        state = current.copy(progress = message.progress)
    }

    private fun applyResultUpserted(message: NodeScannerReconcileMessage.ResultUpserted) {
        state =
            updatingResults(message.generation) { results ->
                val index = results.indexOfFirst { resultIp(it) == resultIp(message.result) }

                if (index >= 0) {
                    results[index] = message.result
                } else {
                    results.add(message.result)
                }
            } ?: return
    }

    private fun applyResultRemoved(message: NodeScannerReconcileMessage.ResultRemoved) {
        state =
            updatingResults(message.generation) { results ->
                results.removeAll { resultIp(it) == message.ip }
            } ?: return
    }

    private fun applyScanStopped(message: NodeScannerReconcileMessage.ScanStopped) {
        val current = scanningState(message.generation) ?: return

        state =
            NodeScannerState.Stopped(
                generation = current.generation,
                network = current.network,
                results = current.results,
            )
    }

    private fun applyScanFinished(message: NodeScannerReconcileMessage.ScanFinished) {
        val current = scanningState(message.generation) ?: return

        state =
            NodeScannerState.Finished(
                generation = current.generation,
                network = current.network,
                results = current.results,
            )
    }

    private fun applyScanFailed(message: NodeScannerReconcileMessage.ScanFailed) {
        val current = scanningState(message.generation) ?: return

        state =
            NodeScannerState.Failed(
                generation = current.generation,
                network = current.network,
                error = message.error,
                results = current.results,
            )
    }

    private fun scanningState(generation: ULong): NodeScannerState.Scanning? =
        (state as? NodeScannerState.Scanning)?.takeIf { it.generation == generation }

    private val currentGeneration: ULong?
        get() =
            when (val current = state) {
                NodeScannerState.Idle -> null
                is NodeScannerState.Scanning -> current.generation
                is NodeScannerState.Stopped -> current.generation
                is NodeScannerState.Finished -> current.generation
                is NodeScannerState.Failed -> current.generation
            }

    private fun updatingResults(
        generation: ULong,
        update: (MutableList<NodeScannerResult>) -> Unit,
    ): NodeScannerState? =
        when (val current = state) {
            NodeScannerState.Idle,
            is NodeScannerState.Failed,
            -> {
                null
            }

            is NodeScannerState.Scanning -> {
                if (current.generation != generation) return null

                current.copy(results = current.results.updated(update))
            }

            is NodeScannerState.Stopped -> {
                if (current.generation != generation) return null

                current.copy(results = current.results.updated(update))
            }

            is NodeScannerState.Finished -> {
                if (current.generation != generation) return null

                current.copy(results = current.results.updated(update))
            }
        }

    private fun List<NodeScannerResult>.updated(
        update: (MutableList<NodeScannerResult>) -> Unit,
    ): List<NodeScannerResult> = toMutableList().also(update)

    private fun resultIp(result: NodeScannerResult): String =
        when (result) {
            is NodeScannerResult.Verified -> result.v1.ip
            is NodeScannerResult.TrustRequired -> result.v1.ip
        }

    override fun reconcile(message: NodeScannerReconcileMessage) {
        if (isClosed.get()) return

        mainScope.launch {
            apply(message)
        }
    }

    override fun reconcileMany(messages: List<NodeScannerReconcileMessage>) {
        if (isClosed.get()) return

        mainScope.launch {
            messages.forEach(::apply)
        }
    }

    @Synchronized
    override fun close() {
        if (rustGuard.isClosed()) return

        mainScope.coroutineContext.cancelChildren()

        try {
            runBlocking {
                withRustOrSuspend(Unit) {
                    stopScan()
                }
            }
        } catch (error: IllegalStateException) {
            Log.w(TAG, "Failed to stop node scan while closing", error)
        } finally {
            rustGuard.closeOnce {
                rust.close()
            }
            mainScope.cancel()
        }
    }

    private companion object {
        const val TAG = "NodeScannerManager"
    }
}
